### PR TITLE
Build shadow once for extra tests

### PR DESCRIPTION
**Pull request by sporksmith**
Previously we were building shadow for each version of tor that we test, as well as to test examples and do bandwidth tests with tgen. These all use the same build configuration and container, so we can save some GH resources by putting these all in one workflow and building shadow once.

This also updates shadow configs and docs to use PATH-based search to locate tgen and tor. It's not strictly necessary for this change, but helps to decouple the "build tgen" job a bit from the other jobs that use its output, and is something we're moving towards in general to support alternative install locations.

### CODE REVIEW

**Comment by sporksmith** [comment]
Here's an example of the full workflow: https://github.com/shadow/shadow/actions/runs/3574968073

It'd be nice to reduce the shadow build time too, but at least after this change we're building it once instead of 5x.

image

**Review by stevenengler — APPROVED** [review]

File: .github/workflows/run_tor.yml [hunk 1]
```diff
@@ -16,37 +16,22 @@ on:
...
-  tor:
-    # use the oldest kernel supported by github's CI (make sure to update the
-    # minimum supported kernel version in documentation when changing)
-    # https://github.com/actions/virtual-environments
+  build_shadow:
     runs-on: ubuntu-20.04
-
     container:
-      image: ${{ matrix.container }}
+      image: 'ubuntu:22.04'
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply .github/workflows/run_tor.yml]: Yeah, I'd tried that, but sadly it doesn't work and there doesn't seem to be a good way to have workflow-level variables that can be used outside of shell snippets. e.g. https://github.com/orgs/community/discussions/26671

Added comments.

File: .github/workflows/run_tor.yml [hunk 2]
```diff
@@ -16,37 +16,22 @@ on:
...
       # the default shm-size for ubuntu:18.04, but with the size increased from
       # 65536k. github's default docker seccomp policy seems to disallow
       # process_vm_readv and process_vm_writev; disable it altogether. See
       # https://docs.docker.com/engine/security/seccomp/
       options: '--tmpfs /dev/shm:rw,nosuid,nodev,exec,size=1024g --security-opt seccomp=unconfined'
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply .github/workflows/run_tor.yml]: Done

File: .github/workflows/run_tor.yml [hunk 3]
```diff
@@ -103,46 +64,201 @@ jobs:
...
+    runs-on: ubuntu-20.04
+    container:
+      image: 'ubuntu:22.04'
+    steps:
+      - name: Restore cache
+        id: restore-tgen-build-cache
+        uses: actions/cache@v3
+        with:
+          path: opt/tgen
+          key: tgen-build-key-7a5cf4554f813cc996637d942f169aeb745e37ff
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply .github/workflows/run_tor.yml]: done

File: .github/workflows/run_tor.yml [hunk 4]
```diff
@@ -103,46 +64,201 @@ jobs:
...
+            libigraph-dev
 
       - name: Restore tor build cache
         id: restore-tor-build-cache
         uses: actions/cache@v3
         with:
-          path: tor
-          key: tor-build-${{ matrix.container }}-${{ matrix.tor }}
+          path: opt/tor
+          key: tor-build-keyv3-${{ matrix.tor }}
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply .github/workflows/run_tor.yml]: done

File: src/test/tor/minimal/tor-minimal.yaml [hunk 5]
```diff
@@ -23,100 +23,100 @@ hosts:
   fileserver:
     network_node_id: 0
     processes:
-    - path: ~/.local/bin/tgen
+    - path: tgen
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/test/tor/minimal/tor-minimal.yaml:26]: done

File: .github/workflows/run_tor.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (sporksmith) [reply .github/workflows/run_tor.yml]: Actually it shows `env` as available from `container`, but not from `container.image`. Weird.

Seeing if I can at least use `${{ job.container }}` in the cache keys...

**Pull request merged**